### PR TITLE
Fix #499 anf #500

### LINKED ISSUE #499: Param flags is ignored in File.open
```ruby
      File.open(
        fname,
        flags: File::WRONLY | File::EXCL | File::CREAT,
        external_encoding: Encoding::BINARY,
      )
```
works fine in the real world, fails in tests with fakefs

```
Errno::ENOENT: No such file or directory - path/filename.gz
from /home/vladimir/.local/share/rtx/installs/ruby/2.3.8/lib/ruby/gems/2.3.0/gems/fakefs-1.2.1/lib/fakefs/file.rb:873:in `check_file_existence!'
```

If I replace `flags` with `mode`, it works fine

### PR DESCRIPTION
**Pull request by grosser**

### CODE REVIEW

File: lib/fakefs/file.rb [hunk 1]
```diff
@@ -490,17 +506,78 @@ def check_if_bit_set(bit)
...
+
+      # truncate doesn't work
+      @file.content.force_encoding(Encoding.default_external)
+      super(@file.content, mode, **opts)
+
+      # super('', mode, **opts)
+      # ext_enc = self.external_encoding
+      # # because it is explicitly removed
+      # StringIO.instance_method(:string=).bind(self).call(@file.content)
+      # self.set_encoding(ext_enc)
```
Comment (grosser) [inline lib/fakefs/file.rb:575]: what is with this code ?

File: lib/fakefs/file.rb [hunk 2]
```diff
@@ -490,17 +506,78 @@ def check_if_bit_set(bit)
...
+          @oflags |= intflags
+          @fmode = create_fmode(@oflags)
+        else
+          raise TypeError, "no implicit conversion of #{opts[:flags].class} into Integer"
+        end
+      end
+      @fmode ||= create_fmode(@oflags)
+
+      # TODO: extract_binmode && rb_io_ext_int_to_encs, parse_mode_enc
+      # UPD: Looks like we don't need that with our pretty StringIO hack
```
Comment (grosser) [inline lib/fakefs/file.rb:556]: so can remove ?

File: lib/fakefs/file.rb [hunk 3]
```diff
@@ -12,7 +12,24 @@ class File < StringIO
...
+    FMODE_CREATE = 0x00000080
+    FMODE_EXCL = 0x00000400
+    FMODE_TRUNC = 0x00000800
+    FMODE_TEXTMODE = 0x00001000
+
+    TEXT_MODES = {
+      'r' => FMODE_READABLE,
+      'w' => FMODE_WRITABLE | FMODE_TRUNC | FMODE_CREATE,
+      'a' => FMODE_WRITABLE | FMODE_APPEND | FMODE_CREATE
+    }.freeze
```
Comment (grosser) [inline lib/fakefs/file.rb:29]: where did you get these from ... some internal ruby place we can reference directly instead of copying ?

**Comment by grosser** [comment]
@uvlad7 can you check comments

File: lib/fakefs/file.rb (hunk 2 above)
Comment (uvlad7) [reply lib/fakefs/file.rb]: yep, sorry, forgot to do that

**Comment by grosser** [comment]
first run failed on some random error, but fine after retry

**Comment by grosser** [comment]
this reproduces the CI error, do you know why that would happen ?
```
bundle exec mtest test --seed 42242
```
... works fine on master

File: lib/fakefs/file.rb (hunk 3 above)
Comment (uvlad7) [reply lib/fakefs/file.rb:29]: I copied them from the ruby [header](https://github.com/ruby/ruby/blob/v2_7_8/include/ruby/io.h#L108). I doubt it's required to access them directly, these - unlike `O_SOMETHING` constants like `O_WRONLY`, aka `File::WRONLY` - are just bit flags well crafted for the purpose.
BTW, it looks like File constants are unnecessary and can be replaced with `FMODE`-s (but you cannot use `File::BINARY`, for example, it's just zero on linux, and ruby uses `FMODE_BINMODE` in `binmode?`)

Comment (grosser) [reply lib/fakefs/file.rb:29]: kk, added a inline comment

**Comment by uvlad7** [comment]
> this reproduces the CI error, do you know why that would happen ?
> 
> ```
> bundle exec mtest test --seed 42242
> ```
> 
> ... works fine on master

Oh, I faced this - EISDIR - during development, but I thought it was because other test failed and `io_test.rb` doesn't perform a cleanup when fails. But now I see only these tests failed

**Comment by uvlad7** [comment]
I added ensure clauses everywhere in `io_test`, `test_method_IO_binread_works` still fails, lookin into that
```
    ensure
      ::FakeFS.deactivate!
      ::FakeFS::FileSystem.clear
    end
```

**Comment by uvlad7** [comment]
Here we are, still some cleanup
```
From: /home/vladimir/fakefs/lib/fakefs/file.rb:521 FakeFS::File#initialize:

    516:       end
    517:       mode, _perm = args
    518: 
    519:       @path = path
    520:       @file = FileSystem.find(@path)
 => 521:       binding.pry if $pry_please
    522:       # real rb_scan_open_args - and rb_io_extract_modeenc - is much more complex
    523:       raise ArgumentError, 'mode specified twice' unless mode.nil? || opts[:mode].nil?
    524: 
    525:       mode_opt = mode.nil? ? opts[:mode] : mode
    526:       # see vmode_handle

[1] pry(#<FakeFS::File>)> @file
=> (FakeDir name:"foo" parent:"/" size:2)
[2] pry(#<FakeFS::File>)> @path
=> "foo"
```

Don't know how to see the prev test name

**Comment by uvlad7** [comment]
@grosser just pushed a fix, not sure what to do with `test/test_helper.rb`. I suppose it's used outside the tests, aka a part of the gem public interface?

**Comment by grosser** [comment]
test_helper is not used outside of tests

**Comment by grosser** [comment]
picked the commit here since original PR had merge conflicts

**Comment by uvlad7** [comment]
Looks like this one [commit](https://github.com/uvlad7/fakefs/commit/3e89db275194393332cd00399098fde781527454) was lost. Merged upstream master and created a PR without conflicts) #508 
And fixed my name in `CONTRIBUTORS`

**Pull request merged**